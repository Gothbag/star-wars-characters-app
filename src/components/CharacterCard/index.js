import React, {Fragment} from "react";
import PropTypes from "prop-types";
import { isEmpty, noop } from "lodash";

import "./CharacterCard.css";

import { getIdFromUrl } from "../../helpers/getHelpers";

class CharacterCard extends React.PureComponent {

    static propTypes = {
        films: PropTypes.objectOf(PropTypes.string),
        character: PropTypes.shape({
        	birthYear: PropTypes.string,
        	gender: PropTypes.string,
            name: PropTypes.string.isRequired
        }),
        loadPlanet: PropTypes.func,
        loadSpecies: PropTypes.func,
        planets: PropTypes.objectOf(PropTypes.string),
        species: PropTypes.objectOf(PropTypes.string)
    }

    static defaultProps = {
        films: {},
    	character: {},
    	loadPlanet: noop,
      	loadSpecies: noop,
        planets: {},
        species: {}
    }

    componentDidUpdate() {
        const { props: { character, planets, species, loadPlanet, loadSpecies } } = this;
        if (!isEmpty(character) && !(species && species[character.species])) {
        	loadSpecies(character.species);
        }
        if (!isEmpty(character) && !(planets && planets[getIdFromUrl(character.homeworld)])) {
        	loadPlanet(character.homeworld);
        }
    }

    getFilmStr = films => Array.isArray(films) ? films.map(film => this.props.films[film]).join(", ") : [];

    render() {
        const { getFilmStr, props: { character: { birthYear, films, gender, homeworld, name, species: speciesId }, planets, species } } = this;
        return (<div className="panel panel-default">
            <div className="panel-heading">{name || "Search for a character"}</div>
            {!isEmpty(this.props.character) && <Fragment>
                <div className="panel-body">
                    <p>Birth year: {birthYear}</p>
                    {planets[homeworld] && <p>Homeworld: {planets[homeworld]}</p>}
                    {species[speciesId] && <p>Species: {species[speciesId]}</p>}
                    <p>Gender: {gender}</p>
                    <p></p>
                </div>
                <div className="panel-footer">Films: {getFilmStr(films)}</div>
            </Fragment>}
        </div>);
    }
}

export default CharacterCard;
