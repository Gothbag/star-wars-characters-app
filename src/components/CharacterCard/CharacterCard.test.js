import React from 'react';
import CharacterCard from './index';

describe("<CharacterCard/>", () => {

	const props = {

	};

	beforeEach(() => {
		props.loadSpecies = jest.fn();
		props.loadPlanet = jest.fn();
		props.planets = {};
		props.species = {};
		props.character = {};
	});

	it('renders without crashing', () => {
		const wrapper = shallow(<CharacterCard/>);
		expect(wrapper).toMatchSnapshot();
	});

	it("should call the loadPlanet and loadSpecies props if the character is not empty", () => {
		const wrapper = mount(<CharacterCard {...props}/>);
		wrapper.setProps({
			character: { name: "John" }
		});
		expect(props.loadSpecies.mock.calls.length).toBe(1);
		expect(props.loadPlanet.mock.calls.length).toBe(1);
	});

	it("shouldn't call the loadPlanet and loadSpecies props if the planets and species props have matching ids", () => {
		props.character = { name: "John", species: "1", homeworld: "1" };
		props.planets = {
			1: "Tatooine"
		};
		props.species = {
			1: "Human"
		};
		const wrapper = shallow(<CharacterCard {...props}/>);
		expect(props.loadSpecies.mock.calls.length).toBe(0);
		expect(props.loadPlanet.mock.calls.length).toBe(0);
	});

});
