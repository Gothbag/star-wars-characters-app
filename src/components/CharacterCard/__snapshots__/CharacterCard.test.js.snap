// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CharacterCard/> renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CharacterCard
    character={Object {}}
    films={Object {}}
    loadPlanet={[Function]}
    loadSpecies={[Function]}
    planets={Object {}}
    species={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="panel-heading"
>
          Search for a character
</div>,
        false,
      ],
      "className": "panel panel-default",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Search for a character",
          "className": "panel-heading",
        },
        "ref": null,
        "rendered": "Search for a character",
        "type": "div",
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="panel-heading"
>
            Search for a character
</div>,
          false,
        ],
        "className": "panel panel-default",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Search for a character",
            "className": "panel-heading",
          },
          "ref": null,
          "rendered": "Search for a character",
          "type": "div",
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`<CharacterCard/> should call the loadPlanet and loadSpecies props if the character is not empty 1`] = `0`;

exports[`<CharacterCard/> should call the loadPlanet and loadSpecies props if the character is not empty 2`] = `0`;

exports[`<CharacterCard/> shouldn't call the loadPlanet and loadSpecies props if matching 1`] = `0`;

exports[`<CharacterCard/> shouldn't call the loadPlanet and loadSpecies props if matching 2`] = `0`;

exports[`<CharacterCard/> shouldn't call the loadPlanet and loadSpecies props if the planets and species props have matching ids 1`] = `0`;

exports[`<CharacterCard/> shouldn't call the loadPlanet and loadSpecies props if the planets and species props have matching ids 2`] = `0`;

exports[`renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CharacterCard
    character={Object {}}
    films={Object {}}
    loadPlanet={[Function]}
    loadSpecies={[Function]}
    planets={Object {}}
    species={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="panel-heading"
>
          Search for a character
</div>,
        false,
      ],
      "className": "panel panel-default",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Search for a character",
          "className": "panel-heading",
        },
        "ref": null,
        "rendered": "Search for a character",
        "type": "div",
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="panel-heading"
>
            Search for a character
</div>,
          false,
        ],
        "className": "panel panel-default",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Search for a character",
            "className": "panel-heading",
          },
          "ref": null,
          "rendered": "Search for a character",
          "type": "div",
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
